# Introduction to pandas

---

In this notebook, we will learn the basics of pandas, a commonly used data science library

### Table of Contents

1 - [Introduction](#section intro)<br>

2 - [DataFrames](#section dataframes)<br>

3 - [Tools for Examining DataFrames](#section tools)<br>

4 - [Indexing into DataFrames](#section indexing)<br>

5 - [Manipulating DataFrames](#section manipulating)<br>

6 - [Wrap-Up](# section wrapup)<br>

Estimated time: 30 minutes

---

## 1. Introduction <a id='intro'></a>

<i>What is pandas?</i>

Here is the Wikipedia definition of pandas: "pandas is a software library written for the Python programming language for data manipulation and analysis. In particular, it offers data structures and operations for manipulating numerical tables and time series."

Basically, pandas is yet another tool we can use to manipulate data. 

Let's begin our exploration of pandas! First, we need to import it. The standard way of importing pandas is with this line of code:

In [2]:
import pandas as pd

---

## 2. DataFrames <a id='dataframes'></a>

A data structure is a particular way of representing and storing data in code. The <b>DataFrame</b> is the most used data structure in pandas. 

<b>Definition</b>: A DataFrame is a 2-dimensional labeled data structure with columns of potentially different types.

Basically, a DataFrame is a table-like thing, with rows and columns, that stores data for us.

Let's create our very first DataFrame. One way we can create a DataFrame is to use pandas' read_csv function, passing in a file name. The data in this file will be turned into a DataFrame, which we can save into a variable called df.

In [3]:
df = pd.read_csv("https://raw.githubusercontent.com/vincentarelbundock/Rdatasets/master/csv/datasets/Titanic.csv")

---

## 3. Tools for Examining DataFrames <a id ='tools'></a>

It is always good practice to get some preliminary information about your data before attempting analysis. Here are some pandas functions that will help us do that:

---
a) `head`

The `head` function will give us the first 5 rows in our DataFrame to examine. If we want to get a different number of rows, we can pass in a number when we call the function.

In [4]:
# Gets first 5 rows
df.head()

Unnamed: 0                                           Name PClass    Age  \
0           1                   Allen, Miss Elisabeth Walton    1st  29.00   
1           2                    Allison, Miss Helen Loraine    1st   2.00   
2           3            Allison, Mr Hudson Joshua Creighton    1st  30.00   
3           4  Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00   
4           5                  Allison, Master Hudson Trevor    1st   0.92   

      Sex  Survived  SexCode  
0  female         1        1  
1  female         0        1  
2    male         0        0  
3  female         0        1  
4    male         1        0

In [5]:
# Gets first 10 rows
df.head(10)

Unnamed: 0                                           Name PClass    Age  \
0           1                   Allen, Miss Elisabeth Walton    1st  29.00   
1           2                    Allison, Miss Helen Loraine    1st   2.00   
2           3            Allison, Mr Hudson Joshua Creighton    1st  30.00   
3           4  Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st  25.00   
4           5                  Allison, Master Hudson Trevor    1st   0.92   
5           6                             Anderson, Mr Harry    1st  47.00   
6           7               Andrews, Miss Kornelia Theodosia    1st  63.00   
7           8                         Andrews, Mr Thomas, jr    1st  39.00   
8           9   Appleton, Mrs Edward Dale (Charlotte Lamson)    1st  58.00   
9          10                         Artagaveytia, Mr Ramon    1st  71.00   

      Sex  Survived  SexCode  
0  female         1        1  
1  female         0        1  
2    male         0        0  
3  female         0        1  
4    male         1        0  
5    male         1        0  
6  female         1        1  
7    male         0        0  
8  female         1        1  
9    male         0        0

<b>Question 3.1:</b> How do you get the first 3 rows?

In [ ]:
### YOUR CODE HERE

---

b) `tail`

The `tail` function can give us the last rows of the table.

In [6]:
# Gets last 5 rows
df.tail()

Unnamed: 0                    Name PClass   Age   Sex  Survived  SexCode
1308        1309      Zakarian, Mr Artun    3rd  27.0  male         0        0
1309        1310  Zakarian, Mr Maprieder    3rd  26.0  male         0        0
1310        1311        Zenni, Mr Philip    3rd  22.0  male         0        0
1311        1312        Lievens, Mr Rene    3rd  24.0  male         0        0
1312        1313          Zimmerman, Leo    3rd  29.0  male         0        0

<b>Question 3.2</b> How do you think we can get the last 3 rows?

In [8]:
### YOUR CODE HERE

---

c) `columns`

We can look at all of the column names of our DataFrame by using the `columns` attribute of our DataFrame

In [12]:
# Note that we are not using parentheses. This is because columns is an attribute of df, not a function
df.columns

Index(['Unnamed: 0', 'Name', 'PClass', 'Age', 'Sex', 'Survived', 'SexCode'], dtype='object')

---

d) `describe`

We can use the `describe` function to get a statistic summary of our data.

In [11]:
df.describe()

Unnamed: 0         Age     Survived      SexCode
count  1313.000000  756.000000  1313.000000  1313.000000
mean    657.000000   30.397989     0.342727     0.351866
std     379.174762   14.259049     0.474802     0.477734
min       1.000000    0.170000     0.000000     0.000000
25%     329.000000   21.000000     0.000000     0.000000
50%     657.000000   28.000000     0.000000     0.000000
75%     985.000000   39.000000     1.000000     1.000000
max    1313.000000   71.000000     1.000000     1.000000

<b>Question 3.3:</b> Looking at our describe table, what was the mean age and standard deviation (in years) of passengers on the Titanic?

Mean (in years):

Standard Deviation (in years):


## 4. Indexing into DataFrames <a id='indexing'></a>

This next section will discuss how to actually access the values in our DataFrame. We call this <i>indexing</i>.

---

a) Selecting columns

In order to get a column of our DataFrame, we use bracket notation with the column name. <br><br>

<center>`data_frame_name[column_name]`</center>

Here's how we would get the Age column out of our DataFrame:

In [16]:
df['Age']

0       29.00
1        2.00
2       30.00
3       25.00
4        0.92
5       47.00
6       63.00
7       39.00
8       58.00
9       71.00
10      47.00
11      19.00
12        NaN
13        NaN
14        NaN
15      50.00
16      24.00
17      36.00
18      37.00
19      47.00
20      26.00
21      25.00
22      25.00
23      19.00
24      28.00
25      45.00
26      39.00
27      30.00
28      58.00
29        NaN
        ...  
1283    14.00
1284    22.00
1285      NaN
1286      NaN
1287      NaN
1288      NaN
1289      NaN
1290      NaN
1291    51.00
1292    18.00
1293    45.00
1294      NaN
1295      NaN
1296      NaN
1297    28.00
1298    21.00
1299    27.00
1300      NaN
1301    36.00
1302      NaN
1303    27.00
1304    15.00
1305      NaN
1306      NaN
1307      NaN
1308    27.00
1309    26.00
1310    22.00
1311    24.00
1312    29.00
Name: Age, Length: 1313, dtype: float64

Notice that the column names are case sensitive! 

<b>Question 4.1:</b> What happens when you try to access the `age` column from our DataFrame?

In [18]:
### Try it out!

<b>Answer:</b>

<b>Question 4.2</b>: Index into our DataFrame to get the `Survived` column!

In [ ]:
### YOUR CODE HERE

---

b) Selecting Rows

In order to select rows from our DataFrame, we use square brackets and specify the range of rows we want.

For example,

<center>`df[0:3]`</center>

would give us the first three rows of our DataFrame.

The range of rows we get back is inclusive of the first number, but not of the second. In the example above, this means that we get the 0th row, the 1st row, and the 2nd row, but <b>not</b> the 3rd row. 

Does that make sense?

<b>Question 4.3:</b> Get the 100th row to 103rd row of our DataFrame, including the 103rd row.

Hint: You should have 4 rows in your result.

In [19]:
### YOUR CODE HERE

---

c) Selecting a subset of the columns and the rows

We've learned so far how to get out certain columns and rows of our DataFrame, but let's say we want a few rows, and a few columns from our DataFrame. How do we do that?

The `iloc` method will do the trick! `iloc` stands for index location, by the way. This is because we are using the index (the numbers) of the rows and the columns to access rows and columns in the DataFrame.

Here's the syntax:

<center>`df.iloc[3:5, 0:2]`</center>

The range before the comma represents the rows we are selecting (in this case, the 3rd row and the 4th row), and the range after the comma represents the columns we are selecting (the 0th column and the 1st column).

Go ahead and run the cell below! Let's see what we get!

In [25]:
df.iloc[3:5, 0:2]

Unnamed: 0                                           Name
3           4  Allison, Mrs Hudson JC (Bessie Waldo Daniels)
4           5                  Allison, Master Hudson Trevor

<b>Question 4.4:</b> Get the 1st row and 2nd and 3rd column out (all of these numbers are by index, remember that Python is 0-indexed)!

In [27]:
### YOUR CODE HERE

## 5. Manipulating DataFrames <a id='manipulating'></a>

Now, we're going to examine a few functions that will enable us to manipulate our DataFrame in various ways.

First, notice that these changes do not persist. Basically, say we have DataFrame d. If we call one of these functions on it, what will be outputted is the result of applying the function on d, but d will remain unchanged.

In order to get our changes to persist, we need to save the result of applying the function into a variable.

---

a) `drop`

The `drop` function allows us to drop rows and columns that are not desired. 

To drop a row we do not want:
<center>`df.drop(1)`</center>

Here, we are dropping the 1st-index row.

If we want to drop a column:
<center>`df.drop(1, axis=1)`</center>

Here, we are dropping the 1st-index column. Note that we have to add axis=1 to drop a column and not a row!


<b>Question 5.1: </b> How would you drop the 27th (indexed column)?

In [32]:
### YOUR CODE HERE

<b>Question 5.2:</b> How would you drop the SexCode column in our DataFrame?

In [33]:
### YOUR CODE HERE

---

b) filtering data with logical subsetting

Let's say you want to get the rows of our DataFrame where the age of the person is older than 30. We would do that with logical subsetting, in which we use a boolean expression to select certain rows. The code would look like:

<center>`df[df['Age'] > 30]`</center>

In [39]:
### Run this line of code! See what happens!
df[df['Age'] > 30]

Unnamed: 0                                               Name PClass  \
5              6                                 Anderson, Mr Harry    1st   
6              7                   Andrews, Miss Kornelia Theodosia    1st   
7              8                             Andrews, Mr Thomas, jr    1st   
8              9       Appleton, Mrs Edward Dale (Charlotte Lamson)    1st   
9             10                             Artagaveytia, Mr Ramon    1st   
10            11                          Astor, Colonel John Jacob    1st   
15            16     Baxter, Mrs James (Helene DeLaudeniere Chaput)    1st   
17            18                                Beattie, Mr Thomson    1st   
18            19                       Beckwith, Mr Richard Leonard    1st   
19            20    Beckwith, Mrs Richard Leonard (Sallie Monypeny)    1st   
25            26                        Blackwell, Mr Stephen Weart    1st   
26            27                                    Blank, Mr Henry    1st   
28            29                            Bonnell, Miss Elizabeth    1st   
30            31                            Bowen, Miss Grace Scott    1st   
33            34                             Brady, Mr John Bertram    1st   
34            35                                  Brandeis, Mr Emil    1st   
36            37   Brown, Mrs James Joseph (Margaret Molly" Tobin)"    1st   
37            38      Brown, Mrs John Murray (Caroline Lane Lamson)    1st   
38            39     Bucknell, Mrs William Robert (Emma Eliza Ward)    1st   
39            40                   Butt, Major Archibald Willingham    1st   
41            42    Candee, Mrs Edward (Helen Churchill Hungerford)    1st   
42            43  Cardeza, Mrs James Warburton Martinez (Charlot...    1st   
43            44                  Cardeza, Mr Thomas Drake Martinez    1st   
44            45                            Carlsson, Mr Frans Olof    1st   
47            48                          Carter, Mr William Ernest    1st   
48            49           Carter, Mrs William Ernest (Lucile Polk)    1st   
51            52                              Case, Mr Howard Brown    1st   
53            54                       Cavendish, Mr Tyrell William    1st   
55            56                         Chaffee, Mr Herbert Fuller    1st   
56            57  Chaffee, Mrs Herbert Fuller (Carrie Constance ...    1st   
...          ...                                                ...    ...   
867          868                   Holm, Mr John Frederik Alexander    3rd   
872          873            Humblin, Mr Adolf Mathias Nicolai Olsen    3rd   
882          883                             Jensen, Mr Niels Peder    3rd   
886          887                          Johanson, Mr Jakob Alfred    3rd   
888          889                         Johansson, Mr Gustaff Joel    3rd   
889          890                           Johansson, Mr Karl Johan    3rd   
892          893                                 Johnson, Mr Alfred    3rd   
894          895                        Johnson, Mr Malkolm Joackim    3rd   
903          904                                   Jonsson, Mr Carl    3rd   
906          907                                   Jussila, Mr Erik    3rd   
915          916                                    Karun, Mr Franz    3rd   
921          922                                    Kelly, Mr James    3rd   
922          923                                    Kelly, Mr James    3rd   
931          932                                     Kink, Mr Anton    3rd   
937          938                         Klasen, Mrs Hulda Kristina    3rd   
942          943                       Laitinen, Miss Kritina Sofia    3rd   
958          959                          Leinonen, Mr Antti Gustaf    3rd   
962          963                                 Leonard, Mr Lionel    3rd   
963          964                                   Lester, Mr James    3rd   
965          966            

<b>Question 5.3:</b> It's your turn! Select only the rows of passengers who survived.

In [40]:
### YOUR CODE HERE

---

c) `dropna`

When dealing with data, we often have to deal with missing values. The `dropna` function drops rows in which there are missing values.

In [44]:
df.dropna()

Unnamed: 0                                              Name PClass  \
0              1                      Allen, Miss Elisabeth Walton    1st   
1              2                       Allison, Miss Helen Loraine    1st   
2              3               Allison, Mr Hudson Joshua Creighton    1st   
3              4     Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st   
4              5                     Allison, Master Hudson Trevor    1st   
5              6                                Anderson, Mr Harry    1st   
6              7                  Andrews, Miss Kornelia Theodosia    1st   
7              8                            Andrews, Mr Thomas, jr    1st   
8              9      Appleton, Mrs Edward Dale (Charlotte Lamson)    1st   
9             10                            Artagaveytia, Mr Ramon    1st   
10            11                         Astor, Colonel John Jacob    1st   
11            12  Astor, Mrs John Jacob (Madeleine Talmadge Force)    1st   
15            16    Baxter, Mrs James (Helene DeLaudeniere Chaput)    1st   
16            17                           Baxter, Mr Quigg Edmond    1st   
17            18                               Beattie, Mr Thomson    1st   
18            19                      Beckwith, Mr Richard Leonard    1st   
19            20   Beckwith, Mrs Richard Leonard (Sallie Monypeny)    1st   
20            21                              Behr, Mr Karl Howell    1st   
21            22                                Birnbaum, Mr Jakob    1st   
22            23                            Bishop, Mr Dickinson H    1st   
23            24            Bishop, Mrs Dickinson H (Helen Walton)    1st   
24            25           Bjornstrm-Steffansson, Mr Mauritz Hakan    1st   
25            26                       Blackwell, Mr Stephen Weart    1st   
26            27                                   Blank, Mr Henry    1st   
27            28                            Bonnell, Miss Caroline    1st   
28            29                           Bonnell, Miss Elizabeth    1st   
30            31                           Bowen, Miss Grace Scott    1st   
31            32                        Bowerman, Miss Elsie Edith    1st   
33            34                            Brady, Mr John Bertram    1st   
34            35                                 Brandeis, Mr Emil    1st   
...          ...                                               ...    ...   
1264        1265                               Turkula, Mrs Hedvig    3rd   
1269        1270                      Van der Planke, Miss Augusta    3rd   
1270        1271                          Van der Planke, Mr Jules    3rd   
1271        1272                         Van der Planke, Mrs Jules    3rd   
1272        1273                           Van der Planke, Mr Leon    3rd   
1273        1274                       Van der Steen, Mr Leo Peter    3rd   
1274        1275                      Van de Velde, Mr John Joseph    3rd   
1275        1276                      Vandewalle, Mr Nestor Cyriel    3rd   
1276        1277                          Van Impe, Miss Catharine    3rd   
1277        1278                        Van Impe, Mr Jean Baptiste    3rd   
1278        1279                       Van Impe, Mrs Jean Baptiste    3rd   
1279        1280                               Vartunian, Mr David    3rd   
1281        1282                             Vendel, Mr Olof Wdvin    3rd   
1282        1283                             Vereruysse, Mr Victor    3rd   
1283        1284               Vestrom, Miss Hulda Amanda Adolfina    3rd   
1284        1285                                    Vonk, Mr Jenko    3rd   
1291        1292                        Widegren, Mr Charles Peter    3rd   
1292        1293                          Wiklund, Mr Jacob Alfred    3rd   
1293        1294                                 Wilkes, Mrs Ellen    3rd   
1297        1298                               Williams, Mr Leslie    3rd   
12

If you look at the number of rows in this DataFrame, there are only 756 rows, when originially, the DataFrame had more than 1000 rows.

---

d) `fillna`

Let's say that instead of getting rid of rows with NAs, we want to fill the NAs with a default value. We can do this with the `fillna` function. All we have to do is call fillna on the DataFrame, passing in what we want to fill in.

In [47]:
df.fillna(3)

Unnamed: 0                                              Name PClass  \
0              1                      Allen, Miss Elisabeth Walton    1st   
1              2                       Allison, Miss Helen Loraine    1st   
2              3               Allison, Mr Hudson Joshua Creighton    1st   
3              4     Allison, Mrs Hudson JC (Bessie Waldo Daniels)    1st   
4              5                     Allison, Master Hudson Trevor    1st   
5              6                                Anderson, Mr Harry    1st   
6              7                  Andrews, Miss Kornelia Theodosia    1st   
7              8                            Andrews, Mr Thomas, jr    1st   
8              9      Appleton, Mrs Edward Dale (Charlotte Lamson)    1st   
9             10                            Artagaveytia, Mr Ramon    1st   
10            11                         Astor, Colonel John Jacob    1st   
11            12  Astor, Mrs John Jacob (Madeleine Talmadge Force)    1st   
12            13                      Aubert, Mrs Leontine Pauline    1st   
13            14                          Barkworth, Mr Algernon H    1st   
14            15                                Baumann, Mr John D    1st   
15            16    Baxter, Mrs James (Helene DeLaudeniere Chaput)    1st   
16            17                           Baxter, Mr Quigg Edmond    1st   
17            18                               Beattie, Mr Thomson    1st   
18            19                      Beckwith, Mr Richard Leonard    1st   
19            20   Beckwith, Mrs Richard Leonard (Sallie Monypeny)    1st   
20            21                              Behr, Mr Karl Howell    1st   
21            22                                Birnbaum, Mr Jakob    1st   
22            23                            Bishop, Mr Dickinson H    1st   
23            24            Bishop, Mrs Dickinson H (Helen Walton)    1st   
24            25           Bjornstrm-Steffansson, Mr Mauritz Hakan    1st   
25            26                       Blackwell, Mr Stephen Weart    1st   
26            27                                   Blank, Mr Henry    1st   
27            28                            Bonnell, Miss Caroline    1st   
28            29                           Bonnell, Miss Elizabeth    1st   
29            30                           Borebank, Mr John James    1st   
...          ...                                               ...    ...   
1283        1284               Vestrom, Miss Hulda Amanda Adolfina    3rd   
1284        1285                                    Vonk, Mr Jenko    3rd   
1285        1286                                Ware, Mr Frederick    3rd   
1286        1287                        Warren, Mr Charles William    3rd   
1287        1288                                  Wazli, Mr Yousif    3rd   
1288        1289                                  Webber, Mr James    3rd   
1289        1290                     Wennerstrom, Mr August Edvard    3rd   
1290        1291                                Wenzel, Mr Linhart    3rd   
1291        1292                        Widegren, Mr Charles Peter    3rd   
1292        1293                          Wiklund, Mr Jacob Alfred    3rd   
1293        1294                                 Wilkes, Mrs Ellen    3rd   
1294        1295                                  Willer, Mr Aaron    3rd   
1295        1296                                 Willey, Mr Edward    3rd   
1296        1297                          Williams, Mr Howard Hugh    3rd   
1297        1298                               Williams, Mr Leslie    3rd   
1298        1299                                Windelov, Mr Einar    3rd   
1299        1300                                   Wirz, Mr Albert    3rd   
1300        1301                             Wiseman, Mr Phillippe    3rd   
1301        1302                           Wittevrongel, Mr Camiel    3rd   
1302        1303                                 Yalsevac, Mr Ivan    3rd   
13

Now, we've replaced all NAs with the number 3!

---

## 6. Wrap-up <a id ='wrapup'></a>

We've learned a lot about pandas in this intro notebook! Here's a summary table of what we've learned. Otherwise: congrats!

|Name|Example|Purpose|
|-|-|-|
|`read_csv`|`read_csv("filename.csv")`|Create a DataFrame with data from `filename.csv`|
|`head`|`df.head(5)`|Get first 5 rows of our DataFrame|
|`tail`|`df.tail(5)`|Get last 5 rows of our DataFrame|
|`columns`|`df.columns`|Get column names in our DataFrame|
|`iloc`|`df.iloc[1:3, 2:4]`|Get certain rows and columns of our DataFrame|
|`drop`|`df.drop(1), df.drop(1, axis=1)`|Drop 1st (indexed) row and drop 1st (indexed) column|
|`dropna`|`df.dropna()`|Drop rows in our DataFrame that have NAs|
|`fillna`|`df.fillna(5)`|Fill NA values with something that you want (5 in this case)|
